bench/bench.ml

```diff
@@ -0,0 +1,480 @@
+
+  type benchmark = with_metrics:bool -> index_v -> unit -> unit
+
+  let add_metrics =
+    let no_tags x = x in
+    fun () -> Metrics.add src no_tags (fun m -> m (Stats.get ()))
+
+  let write ~with_metrics ?(with_flush = false) bindings rw =
+    Array.iter
+      (fun (k, v) ->
+        Index.replace rw k v;
+        if with_flush then Index.flush rw;
+        if with_metrics then add_metrics ())
+      bindings
+
+  let read ~with_metrics bindings r =
+    Array.iter
+      (fun (k, (_ : Context.Value.t)) ->
+        ignore (Index.find r k : Context.Value.t);
+        if with_metrics then add_metrics ())
+      bindings
+
+  let read_absent ~with_metrics bindings r =
+    Array.iter
+      (fun (k, _) ->
+        try ignore (Index.find r k : Context.Value.t)
+        with Not_found -> if with_metrics then add_metrics ())
+      bindings
+
+  let write_random ~with_metrics v () =
+    v ~fresh:true ~readonly:false "" |> write ~with_metrics !bindings_pool
+
+  let write_seq ~with_metrics v =
+    let bindings = Array.copy !bindings_pool in
+    Array.sort (fun a b -> String.compare (fst a) (fst b)) bindings;
+    fun () -> v ~fresh:true ~readonly:false "" |> write ~with_metrics bindings
+
+  let write_seq_hash ~with_metrics v =
+    let hash e = Context.Key.hash (fst e) in
+    let bindings = Array.copy !bindings_pool in
+    Array.sort (fun a b -> Int.compare (hash a) (hash b)) bindings;
+    fun () -> v ~fresh:true ~readonly:false "" |> write ~with_metrics bindings
+
+  let write_rev_seq_hash ~with_metrics v =
+    let hash e = Context.Key.hash (fst e) in
+    let bindings = Array.copy !bindings_pool in
+    Array.sort (fun a b -> -Int.compare (hash a) (hash b)) bindings;
+    fun () -> v ~fresh:true ~readonly:false "" |> write ~with_metrics bindings
+
+  let write_sync ~with_metrics v () =
+    v ~fresh:true ~readonly:false ""
+    |> write ~with_metrics ~with_flush:true !bindings_pool
+
+  let iter ~with_metrics v () =
+    v ~fresh:false ~readonly:true ""
+    |> Index.iter (fun _ _ -> if with_metrics then add_metrics ())
+
+  let find_random ~with_metrics v () =
+    v ~fresh:false ~readonly:false "" |> read ~with_metrics !bindings_pool
+
+  let find_random_ro ~with_metrics v () =
+    v ~fresh:false ~readonly:true "" |> read ~with_metrics !bindings_pool
+
+  let find_absent ~with_metrics v () =
+    v ~fresh:false ~readonly:false ""
+    |> read_absent ~with_metrics !absent_bindings
+
+  let find_absent_ro ~with_metrics v () =
+    v ~fresh:false ~readonly:true ""
+    |> read_absent ~with_metrics !absent_bindings
+
+  let run :
+      with_metrics:bool ->
+      nb_entries:int ->
+      log_size:int ->
+      root:string ->
+      name:string ->
+      benchmark ->
+      Benchmark.result =
+   fun ~with_metrics ~nb_entries ~log_size ~root ~name b ->
+    let indices = ref [] in
+    let index_v ~fresh ~readonly n =
+      let i = Index.v ~fresh ~readonly ~log_size (root // name // n) in
```

Comment:
`n` seems to be "" for all benchmarks, no?

Reply:
Indeed, will move it here instead!

Reply:
Hmm, I'm not sure it _should_ be moved here. 

I think `index_v` should have the same type as `Index.v`; this is just a scoping and finalising mechanism for any indices that are constructed by the benchmarks. 

The fact that each benchmark only constructs a single index (and therefore puts it in `""`) feels like more of a coincidence than anything else. Shouldn't benchmarks be capable of constructing more than one index?

Reply:
That was indeed the motivation for leaking the name argument in my PR, but on the other hand I don't see what benchmark would need to create multiple indices (without being a combination of already existing benchmarks), so I don't mind removing it if that's easier to understand.
There is however an issue with the current implementation: `cleanup` only cleans the `""` index right now, but that's easy to fix.

Reply:
I'll only allow benchs to create one index for now, we can fix it in another PR if needed.